web/src/echo.ts

```diff
@@ -560,6 +561,16 @@ export function process_from_server(messages: ServerMessage[]): ServerMessage[]
                 msg_list.add_messages(msgs_to_rerender_or_add_to_narrow, {});
             }
         }
+
+        for (const msg_list_data of message_lists.non_rendered_data()) {
+            if (!msg_list_data.filter.can_apply_locally()) {
+                // Ideally we would ask server to if messages matches filter
+                // but it is not worth doing so for every new message.
+                message_list_data_cache.remove(msg_list_data.filter);
```

Comment:
This is probably a case where we could refine the logic to be more precise -- if other operators on the `filter` mean the message can't match, we could not discard. And we could potentially extend the `MessageListData` structure with a list of messages to check before rendering it.

I'll open a follow-up issue, since most searches will just have the keyword operator anyway.

Reply:
Yeah, makes sense.